external_libraries/amgcl/backend/builtin.hpp

```diff
@@ -1029,11 +1029,41 @@ struct inner_product_impl<
 
     typedef typename math::inner_product_impl<V>::return_type return_type;
 
-    static return_type get(const Vec1 &x, const Vec2 &y)
-    {
+    static return_type get(const Vec1 &x, const Vec2 &y) {
+#ifdef _OPENMP
+        if (omp_get_max_threads() > 1) {
+            return parallel(x, y);
+        } else
+#endif
+        {
+            return serial(x, y);
+        }
+    }
+
+    static return_type serial(const Vec1 &x, const Vec2 &y) {
         const size_t n = x.size();
+
+        return_type s = math::zero<return_type>();
+        return_type c = math::zero<return_type>();
+
+        for(ptrdiff_t i = 0; i < static_cast<ptrdiff_t>(n); ++i) {
+            return_type d = math::inner_product(x[i], y[i]) - c;
+            return_type t = s + d;
+            c = (t - s) - d;
+            s = t;
+        }
+
+        return s;
+    }
+
 #ifdef _OPENMP
-        return_type              _sum_stat[64];
+#  ifndef AMGCL_MAX_OPENMP_THREADS
+#    define AMGCL_MAX_OPENMP_THREADS 64
```

Comment:
are u sure of this ? i guess you should check somewhere that this is enough

Reply:
This was already set to 64. This change simply allows to customize the value with a preprocessor definition.

Reply:
Also this is not a hard limit, the inner product will work with larger number of cores, but will use more expensive dynamic vector in that case.